settings.txt
Remember: only *change* numbers, do not add or remove them. 
These settings are fail-safe settings.
You might want to change it in certain circumstances.
AND DO NOT write 'Zero' or 'One'! Insert instead the actual number ;-).
(yes, I know, I should work over the file input algorithm :} )
Oh and please do not hesitate to ask me questions, I will put them in the FAQ :)

---------------
Input / Output:
---------------

Colors 1
-- You might want to turn this to 'Zero' if your screen does not support
-- colors.

Detailed Output 1
-- a 'One' means that the build order in the output file 'bo_*.txt' will be
-- printed with a timestamp [min:sec] in front of each order.
-- a 'Zero' will result in a compact view of the build order which is probably
-- much more readable

Line correction 0
-- Some consoles have an additional line (twenty five instead of twenty four).
-- If you set this to 'Zero' it looks better in some consoles (like the
-- standard linux text console)

Gizmo 1
-- That scrolling text at the top and at the bottom :)
-- Just set it to 'Zero' if you are sick of it ;-)
-- I will work over the text in the next version

------------
Map Options:
------------

Time to Enemy 0
-- This is the time in seconds a 'normal' unit will need to reach the base of
-- the enemy. This setting is especially useful when trying a rush.
-- SCC will adjust the time to each unit as the speed is different.
-- I will add here another setting for air units later.
-- A zero will switch this option off.
-- WARNING: IF YOU SET BOTH 'Time to Enemy' AND THE TIMES FOR UNITS IN THE 
-- 'goal_*.txt' FILES, RETHINK WHETHER THIS GOALS ARE POSSIBLE...
-- example: 'Time to Enemy' is set to ninty while two zealots are set 
-- to five minutes: ninty seconds (ToT) + eighty (two zealots) + thirty
-- (pylon) + sixty (gateway) = ~four minutes... Theoretically enough
-- but you also have to gather minerals. The calculation will end in 
-- a weird result... :[

Vespine Geysirs 1
-- This setting describes how many Vespine Geysirs are near the main base.
-- As there are no rules for an expansion in the calculator this really
-- only means how many of them are near your starting position.

Mineral Gathering modificator 100 
-- Well... I still have a hard time to figure out how much a drone/scv/probe
-- really gathers. I did some tests, stopped the time etc. I think the
-- relations between the number of gatherers is more or less correct.
-- Primarily this describes the distance CC/Nexus/Hatch to the Mineral blocks

Mineral Blocks 8
-- If there are more workers than blocks then some workers have to wait.
-- i.e. if you set Mineral blocks to for example sixteen every single worker
-- will gather the same amount of minerals / minute no matter whether you have
-- four, eight or sixteen.

Time to wallin 0 *not yet implemented*
-- Should defensive buildings be placed away from the main building?
-- This result in higher delays as the drone has to reach the building site
-- first.
-- A 'Zero' deactivates this function.

Scout Time 0
-- At which time a probe / scv should be dispatched to scout the enemy.
-- In program mechanics this means: This probe/scv is lost after 'Scout Time'
-- seconds. Set it to 'Zero' to deactivate it.
-- It is deactivated for the Zerg.
-- Note: The scout probe still needs supply and it is not mentioned in the
-- build order directly

-----------------
Internal Options:
-----------------
Generate Build order 0

Max Time 30 (in minutes)
-- Max Time is to define the search space at the beginning.
-- That means: The program will run slow at the beginning when using a high
-- value and visa versa it will run fast at the beginning when using a small
-- value. 
-- Besides that a high value will grant that the program will find a basic 
-- solution (which fulfills the goals stated in goals_*.txt)
-- A good thumb rule: Set it to thrice the value you would need to build it.

Max Generations 200
-- Max Generations define the time how long a build order should be optimized.
-- After *Max Generations* Generations without change in time and gathered
-- resources, the current run will end and saved.

Mutations 20 
-- How many items in a build order of each test case should have the chance to
-- mutate? Very small values will result in a slow evolution where only small
-- changes per step are possible. It might remain in a local optima :/
-- On the other side very large values will result mainly in destructive
-- mutations which will slow down evolution too... 
-- Fifty worked quite well though there is probably a better number depending
-- on the task

Mutation Rate 100
-- Mutation rate is the probability of a change in the build order.
-- In contrary to above, a large value means that there is a low probability
-- of a mutation and visa versa a small value will result in a high
-- probability.

Verbose Messages 1
-- You probably want to let this at its default setting...
-- Setting it to 'Zero' will cause the program to just output the file and
-- that's about it. NO Debug messages etc.


Did you miss something? Write me!
ghoul@clawsoftware.de


src/settings.txt
Remember: only *change* numbers, do not add or remove them. 
These settings are fail-safe settings.
You might want to change it in certain circumstances.
AND DO NOT write 'Zero' or 'One'! Insert instead the actual number ;-).
(yes, I know, I should work over the file input algorithm :} )
Oh and please do not hesitate to ask me questions, I will put them in the FAQ :)

---------------
Input / Output:
---------------

Colors 1
-- You might want to turn this to 'Zero' if your screen does not support
-- colors.

Detailed Output 1
-- a 'One' means that the build order in the output file 'bo_*.txt' will be
-- printed with a timestamp [min:sec] in front of each order.
-- a 'Zero' will result in a compact view of the build order which is probably
-- much more readable

Line correction 0
-- Some consoles have an additional line (twenty five instead of twenty four).
-- If you set this to 'Zero' it looks better in some consoles (like the
-- standard linux text console)

Gizmo 1
-- That scrolling text at the top and at the bottom :)
-- Just set it to 'Zero' if you are sick of it ;-)
-- I will work over the text in the next version

------------
Map Options:
------------

Time to Enemy 0
-- This is the time in seconds a 'normal' unit will need to reach the base of
-- the enemy. This setting is especially useful when trying a rush.
-- SCC will adjust the time to each unit as the speed is different.
-- I will add here another setting for air units later.
-- A zero will switch this option off.
-- WARNING: IF YOU SET BOTH 'Time to Enemy' AND THE TIMES FOR UNITS IN THE 
-- 'goal_*.txt' FILES, RETHINK WHETHER THIS GOALS ARE POSSIBLE...
-- example: 'Time to Enemy' is set to ninty while two zealots are set 
-- to five minutes: ninty seconds (ToT) + eighty (two zealots) + thirty
-- (pylon) + sixty (gateway) = ~four minutes... Theoretically enough
-- but you also have to gather minerals. The calculation will end in 
-- a weird result... :[

Vespine Geysirs 1
-- This setting describes how many Vespine Geysirs are near the main base.
-- As there are no rules for an expansion in the calculator this really
-- only means how many of them are near your starting position.

Mineral Gathering modificator 100 
-- Well... I still have a hard time to figure out how much a drone/scv/probe
-- really gathers. I did some tests, stopped the time etc. I think the
-- relations between the number of gatherers is more or less correct.
-- Primarily this describes the distance CC/Nexus/Hatch to the Mineral blocks

Mineral Blocks 8
-- If there are more workers than blocks then some workers have to wait.
-- i.e. if you set Mineral blocks to for example sixteen every single worker
-- will gather the same amount of minerals / minute no matter whether you have
-- four, eight or sixteen.

Time to wallin 0 *not yet implemented*
-- Should defensive buildings be placed away from the main building?
-- This result in higher delays as the drone has to reach the building site
-- first.
-- A 'Zero' deactivates this function.

Scout Time 0
-- At which time a probe / scv should be dispatched to scout the enemy.
-- In program mechanics this means: This probe/scv is lost after 'Scout Time'
-- seconds. Set it to 'Zero' to deactivate it.
-- It is deactivated for the Zerg.
-- Note: The scout probe still needs supply and it is not mentioned in the
-- build order directly

-----------------
Internal Options:
-----------------

Max Time 30 (in minutes)
-- Max Time is to define the search space at the beginning.
-- That means: The program will run slow at the beginning when using a high
-- value and visa versa it will run fast at the beginning when using a small
-- value. 
-- Besides that a high value will grant that the program will find a basic 
-- solution (which fulfills the goals stated in goals_*.txt)
-- A good thumb rule: Set it to thrice the value you would need to build it.

Max Generations 50
-- Max Generations define the time how long a build order should be optimized.
-- After *Max Generations* Generations without change in time and gathered
-- resources, the current run will end and saved.

Mutations 50 
-- How many items in a build order of each test case should have the chance to
-- mutate? Very small values will result in a slow evolution where only small
-- changes per step are possible. It might remain in a local optima :/
-- On the other side very large values will result mainly in destructive
-- mutations which will slow down evolution too... 
-- Fifty worked quite well though there is probably a better number depending
-- on the task

Mutation Rate 150
-- Mutation rate is the probability of a change in the build order.
-- In contrary to above, a large value means that there is a low probability
-- of a mutation and visa versa a small value will result in a high
-- probability.

Verbose Messages 1
-- You probably want to let this at its default setting...
-- Setting it to 'Zero' will cause the program to just output the file and
-- that's about it. NO Debug messages etc.


Did you miss something? Write me!
ghoul@clawsoftware.de


src/settings.cpp
#include "settings.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "io.h"

#define DATA_SET_SIZE 16

Settings setup;


void Settings::Fatal(char * strn)
{ 
	setColor(31);
	printf("FATAL ERROR: ");
	setColor(37);
	printf("%s\nExiting...",strn);
}

unsigned char Settings::InitSettings()
{
	const unsigned short power[4]={1,10,100,1000};
	unsigned short size,s;
    unsigned char d;
	signed char t;
	unsigned short data[DATA_SET_SIZE];
	char * buffer;


 	for(s=0;s<DATA_SET_SIZE;s++)
		data[s]=0;
 
	
	printf("Checking input file [settings.txt]...\n");
	FILE * pFileS;
        pFileS = fopen ("settings.txt","rb");
	
	if(pFileS==NULL)
	{
		Fatal("Settings file settings.txt not found!");
		return(1);
	}
	printf("File found. Reading...\n");

	fseek (pFileS, 0, SEEK_END);
	size = ftell (pFileS);
	buffer = new char [size];
	fseek (pFileS, 0, SEEK_SET);
	fread(buffer,1,size,pFileS);
	fclose(pFileS);
	printf("File closed, data saved [%i Bytes].\n",size);
	s=0;d=0;
	
	while(s<size)
	{
		if((buffer[s]>=48)&&(buffer[s]<58)&&(d<DATA_SET_SIZE))
		{
			t=0;
			s++;
			while((buffer[s]>=48)&&(buffer[s]<58)&&(s<size))
			{
				t++;s++;
			}
			if(t>3)
			{
				Fatal("Do not use numbers bigger than 9999 in the settings.txt file!");
				delete buffer;
				return(1);
			}
			while(t>=0)
			{
				data[d]+=(buffer[s-t-1]-48)*power[t];
				t--;
			}
			d++;
		}
		s++;
	}

	//TODO: Check for additional numbers and print out an error !

	delete buffer;
	printf("Buffer freed and saved, checking data:\n");
	



	if(d<DATA_SET_SIZE)
	{
		setColor(31);
		printf("Not enough Data sets found (%i out of expected %i).\n",d,DATA_SET_SIZE);
		setColor(37);
		printf("Please report this to ClawSoftware and/or get an original copy of the 'settings.txt'\n");
		printf("Using defaults...\n");
		colors=1;
		Detailed_Output=1;
		Console24Lines=1;
		Gizmo=1;
		Time_to_Enemy=50;
		Vespene_Geysirs=1;
		Mineral_Mod=100;
		Mineral_Blocks=8;
		Time_to_Wallin=0;
		Scout_Time=160;
		GenerateBuildorder=1;
		Max_Time=45;
		Max_Generations=150;
		Mutations=50;
		Mutation_Rate=50;
		Verbose=1;
	}
	else
	{
		colors=data[0];
		Detailed_Output=data[1];
		Console24Lines=data[2];
		Gizmo=data[3];
		Time_to_Enemy=data[4];
		Vespene_Geysirs=data[5];
		Mineral_Mod=data[6];
		Mineral_Blocks=data[7];
		Time_to_Wallin=data[8];
		Scout_Time=data[9];
		GenerateBuildorder=data[10];
		Max_Time=data[11];
		Max_Generations=data[12];
		Mutations=data[13];
		Mutation_Rate=data[14];
		Verbose=data[15];
	}

	

        if((colors!=0)&&(colors!=1))
        {
               printf("ERROR: 'Colors' not correctly initialized [%i].\n",colors);
	       printf("Please use 0 (white/black) or 1 (colors to stress changes in the BO)!\n");
	       printf("Falling back to default: Colors = 1\n");
	       colors=1;
	}
	
	if((Detailed_Output!=0)&&(Detailed_Output!=1))
	 {
	       printf("ERROR: 'Detailed_Output' not correctly initialized [%i].\n",Detailed_Output);
	       printf("Please use 0 (more readable) or 1 (timestamp in front of each order)!\n");
	       printf("Falling back to default: Detailed_Output = 0\n");
	       Detailed_Output=0;
	 }

	if((Console24Lines!=0)&&(Console24Lines!=1))
	{
		printf("ERROR: 'Line correction' is not correctly initialized [%i].\n",Console24Lines);
		printf("Please use 0 (for 25 line consoles like the linux text console) or 1 (standard)!\n");
		printf("Falling back to default: Line_Correction = 1\n");
		Console24Lines=1;
	}

	if((Gizmo!=0)&&(Gizmo!=1))
	{
		printf("ERROR: 'Gizmo' is not correctly initialized [%i].\n",Gizmo);
		printf("Please use 0 (no scrolling text) or 1 (scrolling gizmo text :)!\n");
		printf("Falling back to default: Gizmo = 1\n");
		Gizmo=1;
	}

	if(Time_to_Enemy>300)
	{
		printf("WARNING: 'Time to Enemy' is larger than 5 minutes! [%i]\n",Time_to_Enemy);
		printf("This is not recommended. Try to change the 'settings.txt'\n");	      	
	}

	if(Vespene_Geysirs==0)
	{
		printf("WARNING: 'Vespene Geysirs' is set to 0.\n");
		printf("This is not recommended. No gas can be gathered!\n");
		printf("Try to change the 'settings.txt'\n");
	
	}
	else if(Vespene_Geysirs>4)
	{
		 printf("ERROR: 'Vespene Geysirs' not correctly initialized [%i].\n",Vespene_Geysirs);
                printf("Please use less than 4 Vespene Geysirs!\n");
		printf("Falling back to default: Vespene Geysirs = 1\n");
		Vespene_Geysirs=1;
	}

	if(Mineral_Mod<10)
	{
		printf("WARNING: 'Mineral Gathering modificator' is set below 10.\n");
		printf("This is not recommended. Almost no minerals can be gathered!\n");
		printf("Try to change the 'settings.txt'\n");
	}

	if(Mineral_Blocks==0)
	{
		printf("ERROR: 'Mineral Blocks' is set to zero.\n");
		printf("Are you crazy or what? ;-)\n");
		printf("Falling back to default: Mineral Blocks = 8\n");
	        Mineral_Blocks=8;	
	}
	else if(Mineral_Blocks>31)
	{
		printf("WARNING: 'Mineral Blocks' is set very high! (%i)\n",Mineral_Blocks);
		printf("This is not recommended. Mineral Mining Calculation will be very inaccurate...\n");
		printf("Try to change the 'settings.txt'\n");
	}

	if((GenerateBuildorder!=0)&&(GenerateBuildorder!=1))
	{
	       printf("ERROR: 'GenerateBuildorder' not correctly initialized [%i].\n",GenerateBuildorder);
               printf("Please use 0 (always start from the scratch) or 1 (pregenerated build orders)!\n");
	       printf("Falling back to default: GenerateBuildorder = 1\n");
	                 Verbose=1;
	}
	

	if(Max_Time<5)
	{
		printf("ERROR: 'Max Time' is set too low. [%i]\n",Max_Time);
		printf("Please use at least 10 minutes.\n");
		printf("Falling back to default: Max Time = 45\n");
		Max_Time=45;
	}
	else
	if(Max_Time<10)
	{
		printf("WARNING: 'Max Time' is set very low. [%i]\n",Max_Time);
		printf("The program will probably not work very well.\n");
		printf("Try to change the 'settings.txt'\n");
	} else if(Max_Time>120)
	{
		printf("WARNING: 'Max Time' is set very high. [%i]\n",Max_Time);
		printf("Crazy or what? ;) Well... prepare for a looong run :p\n");
		printf("Try to change the 'settings.txt'\n");		
	}

	if(Scout_Time/60 >Max_Time)
	{
		printf("ERROR: 'Scout Time' is set too high. [%i]\n",Scout_Time);
		printf("Please keep it below 'Max Time'.\n");
		printf("Falling back to default: Scout Time = 160\n");
		Scout_Time=160;
	}

	if(Max_Generations<10)
	{
		printf("ERROR: 'Max Generations' is set too low. [%i]\n",Max_Generations);
		printf("The program is useless with this setting.\n");
		printf("Falling back to default: Max Generations = 150");
		Max_Generations=150;
	}
	else if(Max_Generations<50)
	{
		printf("WARNING: 'Max Generations' is set very low. [%i]\n",Max_Generations);
		printf("It is not very likely that the program will find an optimal solution.\n");
	       printf("Try to change the 'settings.txt'.\n");
	}	       
	
	if(Mutations==0)
	{
		printf("WARNING: 'Mutations' is set to zero.\n");
		printf("This is only useful if you want to test an own build order and determine its speed\n");
		printf("Optimization is set OFF!\n");
	} else
	if(Mutations<10)
	{
		printf("WARNING: 'Mutations' is set very low.\n");
		printf("The program will probably not be able to determine the optimal solution!\n");
		printf("Try to change the 'settings.txt'.\n");
	}
	else 
	if(Mutations>200)
	{
		printf("WARNING: 'Mutations' is set very high. (%i)\n",Mutations);
		printf("The program will probably not be able to determine the optimal solution!\n");
		printf("Try to change the 'settings.txt'.\n");
	}

	if(Mutation_Rate<5)
	{
		printf("ERROR: 'Mutation Rate' is set too low. [%i]\n",Mutation_Rate);
		printf("Please use at least 5! Remember: 1/Mutation Rate = Probability of a Mutation\n");
	        printf("Falling back to default: Mutation Rate = 50");
       		Mutation_Rate=50;	       
	} 
	
	if((Verbose!=0)&&(Verbose!=1))
	{
		printf("ERROR: 'Verbose' not correctly initialized [%i].\n",Verbose);
                printf("Please use 0 (no verbose output) or 1 (verbose output)!\n");
		printf("Falling back to default: Verbose = 1\n");
	        Verbose=1;
	}

	printf("Reformatting Data Minutes -> Seconds\n");
	Max_Time*=60;
	
	setColor(32);
	printf("'settings.txt' successfully checked!\n");
	setColor(37);
	return(0);
};

unsigned char Settings::InitGoal(char Input[11])
{
	const unsigned short power[4]={1,10,100,1000};
        unsigned short size,s,count_goals;
        unsigned char d;
        signed char t;
        unsigned short data[2*MAX_GOALS];
        char * buffer;
	FILE * pFileS;
	char tmp[255];
	
	printf("Moving on to %s...\n",Input);

	for(s=0;s<MAX_GOALS;s++)
	{
		goal[s].what=0;
		goal[s].time=0;
		buildable[s]=0;
		data[2*s]=0;
		data[2*s+1]=0;
	}

	printf("Checking input file [%s]...\n",Input);
	
        pFileS = fopen (Input,"rb");
	if(pFileS==NULL)
	{
		printf("[%s]:",Input);Fatal("Goal file not found!");
		return(1);		
	};
	printf("File found. Reading... ");
	fseek (pFileS, 0, SEEK_END);
	size = ftell (pFileS);
	printf("[%i Bytes]\n",size);
	buffer = new char [size];
	fseek (pFileS, 0, SEEK_SET);
	fread(buffer,1,size,pFileS);
	fclose(pFileS);
	printf("File closed, data saved.\n");
	s=0;d=0;
	
	while(s<size)
	{
		if((buffer[s]>=48)&&(buffer[s]<58))
		{
			t=0;
			s++;
			while((buffer[s]>=48)&&(buffer[s]<58)&&(s<size))
			{
				t++;s++;
			}
			if(t>3)
			{
				printf("[%s]: ",Input);
				Fatal("Do not use numbers bigger than 9999 in the goal file!");
			        delete buffer;
				return(1);
			}
			while(t>=0)
			{
				data[d]+=(buffer[s-t-1]-48)*power[t];
				t--;
			}
			d++;
		}
		s++;
	}
	delete buffer;
	printf("Buffer freed and saved...\n");
	printf("Analyzing data and setting goals...\n");
// TODO: Maybe additional checks of the data sets here...

	count_goals=0;

	for(s=0;s<MAX_GOALS;s++)
	{
		if((stats[race][s].type==3)&&(data[s*2]>1))
		{
			printf("ERROR: Research projects cannot be build more than a single time.\n");
			printf("Section %s in %s: Using 1 instead of %i...\n",stats[race][s].name,Input,data[s*2]);
			data[s*2]=1;
		}
		if((data[s*2+1]>0)&&(data[s*2+1]*60+stats[race][s].BT >= Max_Time))
		{
			printf("ERROR: Build Time nearly exceeds Max Time!\n");
			printf("Try to increase Max Time [%i] in 'settings.txt' or reduce the goal time [%i] in %s!\n",Max_Time/60,data[s*2+1],Input);
			printf("Disabling goal time for %s...\n",stats[race][s].name);
			data[s*2+1]=0;
		}
		goal[s].what=data[s*2];
		goal[s].time=data[s*2+1];

		//TODO: CHECK!
		
		count_goals+=goal[s].what;
	}

	print("Reformatting Data Minutes -> Seconds\n");
	for(s=0;s<MAX_GOALS;s++)
		goal[s].time*=60;
	setColor(32);
	sprintf(tmp,"All goals [%i] successfully initialized!\n\n",count_goals);
	print(tmp);
	setColor(37);
	print("Press ENTER to continue...");
	if(setup.Verbose==1) d=getchar();
	return(0);
}

void Settings::AdjustMining()
{
//	unsigned short i,j,k;
//	float f,g;
//	const double * Harvest_Speed_tmp[RESOURCES][MAX_PEONS];
		// TODO: ARGH hier noch was machen
//	for(j=0;j<RESOURCES;j++) 
//		for(i=0;i<MAX_PEONS;i++) 
//			Harvest_Speed_tmp[j][i]=&HARVEST_SPEED_RACE[race][j][i];
	
	/*g=Mineral_Blocks*Mineral_Mod/800.0;
	for(i=0;i<45;i++)
		if(i*8/Mineral_Blocks<45)
		{
			f=i*(8.0/Mineral_Blocks);
			k=0;
			for(j=0;j<45;j++)
			{
				if((f+1>j)&&(f<=j)) { k=j;j=45;}
			}
			mining[i]=*miningp[k]*g;
		}
	        else mining[i]=*miningp[44]*g;*/
//	for(j=0;j<RESOURCES;j++)
//		for(i=0;i<MAX_PEONS;i++)
//			Harvest_Speed[j][i]=*Harvest_Speed_tmp[j][i];
}

Settings::Settings()
{
	#ifdef WIN32
	scr=CreateConsoleScreenBuffer(GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ | FILE_SHARE_WRITE, 0, CONSOLE_TEXTMODE_BUFFER, 0);
	#endif	
}

Settings::~Settings()
{
	#ifdef WIN32
	CloseHandle(scr);
	#endif	
}
